### DISCUSSION TITLE
Need help to get extra header value from websocket upgrade process.

### DISCUSSION THREAD

**Discussion by calvin2021y** [curl/curl#10470] [Q&A] [answered]
There could be extra header response from server during the upgrade process, for example:

```sh
Access-Control-Allow-Credentials
Access-Control-Allow-Headers
Access-Control-Allow-Origin
Sec-WebSocket-Accept
```

Is there a way to get headers value after the upgrade ?  (not the first request response header,   but the headers send with Sec-WebSocket-Accept)

**Comment by bagder** [comment, marked as answer]
`curl_easy_header()` should offer those headers as well.

**Comment by calvin2021y** [comment]
Thanks for the quick response.

I also want to know how to call this in multi handle with event loop.

I need get this value before any data read/write,  (the server will not send data unless it already read data first), the client can not send data without know the header value.

I don't want to setup a timer to check it(waste time and extra timer clear process),  maybe need add a  new callback type ?

**Reply by bagder** [reply]
Do you want to do WebSockets in a multi handle setup? Then you use `CURLWS_RAW_MODE` and you get the websocket data delivered to the write callback.

You want to *send* websocket data after the switch has succeeded so you want to know when you can do that?

**Reply by calvin2021y** [reply]
Thanks for tips. I will try `CURLWS_RAW_MODE`.

> You want to send websocket data after the switch has succeeded so you want to know when you can do that?

Yes, a callback after server`HTTP/1.1 101 Switching Protocols` response,  and be able to read this response headers. （because I need client send first frame,  and need to know the response headers before send first frame）

In this case the server not send  any data yet, so write callback is not fired yet.

**Reply by bagder** [reply]
Maybe you can figure it out by using [CURLOPT_HEADERFUNCTION](https://curl.se/libcurl/c/CURLOPT_HEADERFUNCTION.html) and trap the end of headers after the 101?

**Reply by calvin2021y** [reply]
thanks for the tips. it work.
